contracts/EVMTokenBag.slb

```diff
@@ -0,0 +1,102 @@
+pragma solidity ^0.5.2;
+pragma experimental ABIEncoderV2;
+
+
+library EVMTokenBag {
+
+    struct Output {
+        address owner;
+        uint valueOrId;
+        bytes32 data;
+        address color;
+    }
+  
+    struct TokenBag {
+        Output[16] bag;
+    }
+    
+    function balanceOf(
+        TokenBag memory self,
+        address color,
+        address owner
+    ) internal pure returns (uint value) {
+        Output memory output;
+        for (uint i = 0; i < self.bag.length; i++) {
+            output = self.bag[i];
+            if (output.owner == owner && output.color == color) {
+      	         value = output.valueOrId;
+      	         break;
+            }
+        }
+    }
+    
+    function readData(
+        TokenBag memory self,
+        address color,
+        uint id
+    ) internal pure returns (bytes32 data) {
+        Output memory output;
+        for (uint i = 0; i < self.bag.length; i++) {
+            output = self.bag[i];
+            if (output.valueOrId == id && output.color == color) {
+                data = output.data;
+                break;
+            }
+        }
+    }
+    
+    struct TransferParams {
+        address color;
+        address from;
+        address to;
+        uint value;
+    }
+    
+    function transfer(
+        TokenBag memory self,
+        TransferParams memory params
+    ) internal pure returns (bool) {
+        Output memory source;
+        Output memory dest;
+        // find source
+        for (uint i = 0; i < self.bag.length; i++) {
+            if (self.bag[i].owner == params.from && self.bag[i].color == params.color) {
+                source = self.bag[i];
+                break;
+            }
+        }
+    
+        // sender does not have enough tokens to send
+        if (source.valueOrId < params.value) return false;
```

Comment:
would this work for data token transfers? if not, let's document the assumptions.

Reply:
http://erc721.org/

ERC721 (and hence ERC1948) do not have a transfer method, only transferFrom. Hence the function signature will be different. Will deal with this problem down the road XD.

Reply:
Since we probably want to have both transfer and transferFrom share most of the implementation.